internal/expr/query.go

```diff
@@ -86,6 +86,73 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 	return &QueryExpr{outputs: pe.outputs, sql: pe.sql, args: qargs}, nil
 }
 
+type AllIter struct {
+	outputs   []field
+	slicePtrs []reflect.Value
+	sliceVals []reflect.Value
+	elemTypes []reflect.Type
+	isPtrElem []bool
+}
+
+func (a *AllIter) NewRow(columns []string) ([]any, error) {
+	var outputVals = []reflect.Value{}
+	for i := range a.sliceVals {
+		newVal := reflect.New(a.elemTypes[i])
+		if !a.isPtrElem[i] {
+			newVal = newVal.Elem()
+		}
+		a.sliceVals[i] = reflect.Append(a.sliceVals[i], newVal)
+		outputVals = append(outputVals, reflect.Indirect(newVal))
+	}
+	return scanArgs(columns, outputVals, a.outputs)
+}
+
+func (a *AllIter) SetSlices() {
+	for i := range a.sliceVals {
+		a.slicePtrs[i].Elem().Set(a.sliceVals[i])
+	}
+}
+
+func (qe *QueryExpr) NewAllIter(slicePointers []any) (*AllIter, error) {
+	var inQuery = make(map[reflect.Type]bool)
+	for _, field := range qe.outputs {
+		inQuery[field.structType] = true
+	}
+
+	var allIter = AllIter{outputs: qe.outputs}
+	for _, ptr := range slicePointers {
+		var isPtrElem = false
+		if ptr == nil {
+			return nil, fmt.Errorf("need pointer to slice, got nil")
+		}
+		ptrVal := reflect.ValueOf(ptr)
+		if ptrVal.Kind() != reflect.Pointer {
+			return nil, fmt.Errorf("need pointer to slice, got %s", ptrVal.Kind())
+		}
+		if ptrVal.IsNil() {
+			return nil, fmt.Errorf("need pointer to slice, got nil pointer")
+		}
+		sliceVal := ptrVal.Elem()
+		sliceType := sliceVal.Type()
+		if sliceType.Kind() != reflect.Slice {
+			return nil, fmt.Errorf("need pointer to slice, got pointer to %s", sliceType.Kind())
+		}
+		elemType := sliceType.Elem()
+		if elemType.Kind() == reflect.Pointer {
+			isPtrElem = true
+			elemType = elemType.Elem()
+		}
+		if !inQuery[elemType] {
+			return nil, fmt.Errorf("type %q does not appear as an output type in the query", elemType.Name())
+		}
```

Comment:
I wonder if we could perform the sequence of operations that we currently do in a single iteration of this for loop within a separate function instead, and then call that function in the loop at each iteration. This function would be something like:

```Go
func generateSliceInfo (ptr any, inQuery map[reflect.Type]bool) sliceInfo {
    // operations done in a single iteration of the current for loop to appear here
}
```
And then we would call it in the for loop as:
```Go
for _, ptr := range slicePointers {
   info := generateSliceInfo(ptr, inQuery)
   allIter.sInfos = append(allIter.sInfos, info)
}
```